Document/0x05e-Testing-Cryptography.md

```diff
@@ -349,36 +265,31 @@ For the security analysis perspective the analysts may perform the following che
+#### Static Analysis
 
-Another API offered by Android is the `KeyChain`, which provides access to private keys and their corresponding certificate chains in credential storage, which is often not used due to the interaction necessary and the shared nature of the Keychain. See the [Developer Documentation](https://developer.android.com/reference/android/security/KeyChain "Keychain") for more details.
+Identify all the instances of symmetric key encryption in code and look for mechanism which loads or provides a symmetric key. You can look for:
 
-A slightly less secure way of storing encryption keys, is in the SharedPreferences of Android. When [SharedPreferences](https://developer.android.com/reference/android/content/SharedPreferences.html "Android SharedPreference API") are initialized in [MODE_PRIVATE](https://developer.android.com/reference/android/content/Context.html#MODE_PRIVATE "MODE_PRIVATE"), the file is only readable by the application that created it. However, on rooted devices any other application with root access can simply read the SharedPreference file of other apps, it does not matter whether `MODE_PRIVATE` has been used or not. This is not the case for the AndroidKeyStore. Since AndroidKeyStore access is managed on kernel level, which needs considerably more work and skill to bypass without the AndroidKeyStore clearing or destroying the keys.
+- symmetric algorithms (like `DES`, `AES`, etc.)
+- specifications for a key generator (like `KeyGenParameterSpec`, `KeyPairGeneratorSpec`, `KeyPairGenerator`, `KeyGenerator`, `KeyProperties`, etc.)
+- classes which uses `java.security.*`, `javax.crypto.*`, `android.security.*` and `android.security.keystore.*` packages.
 
-The last three options are to use hardcoded encryption keys in the source code, having a predictable key derivation function based on stable attributes, and storing generated keys in public places like `/sdcard/`. Obviously, hardcoded encryption keys are not the way to go. This means every instance of the application uses the same encryption key. An attacker needs only to do the work once, to extract the key from the source code - whether stored natively or in Java/Kotlin. Consequently, an attacker can decrypt any other data which was encrypted by the application.
-Next, when you have a predictable key derivation function based on identifiers which are accessible to other applications, the attacker only needs to find the KDF and apply it to the device in order to find the key. Lastly, storing encryption keys publicly also is highly discouraged as other applications can have permission to read the public partition and steal the keys.
+Verify that symmetric keys in all identified instances are not hardcoded. Check if symmetric keys are not:
 
-#### Static Analysis
+- part of application resources
+- part of a static identifier (which might not be clearly within the definition)
+- hardcoded in code
 
```

Comment:
Maybe add something like "part of a derivation function that always ends up with the same result"?

Reply:
I started to consider different angles here. Maybe, I overthink this, but would be better to say that 'value which can be derived from known values' This should cover "part of a derivation function that always ends up with the same result" (which can be a narrow case). 
Example:
* function(application_name)
* function(system_variable)
* function which multiply by 0

What do you think?

Reply:
yes, sounds legit.

Reply:
Corrected.